### PR TITLE
feat(bundler): add support for EV codesigning certs via azuresigntool

### PR DESCRIPTION
**Pull request by tr3ysmith**
This feature adds support for EV code signing via azuresigntool.exe. The original method using using signtool.exe still works. But if the environment variables are set properly before running `tauri build`, the Tauri bundler will attempt to use AzureSignTool.

To use AzureSignTool, the user must have an EV certificate hosted in Azure Key Vault.

### CODE REVIEW

**Comment by tr3ysmith** [comment]
due to the complexity of the azuresigntool package, I removed it from the binary and it'll need to currently be preinstalled on the host machine. With GitHub runners, this should be possible by running  `dotnet tool install --global AzureSignTool` before executing the Tauri build command

**Comment by tr3ysmith** [comment]
@lucasfernog should we bump the rust min version on the main crate too? I had to bump it to 1.72 on the CLI crate because of the dependency for jsonschema -> aHash (aHash requires 1.72)

https://github.com/tkaitchuck/aHash/pull/196

File: tooling/cli/Cargo.toml [hunk 1]
```diff
@@ -6,7 +6,7 @@ name = "tauri-cli"
 version = "1.5.9"
 authors = [ "Tauri Programme within The Commons Conservancy" ]
 edition = "2021"
-rust-version = "1.60"
```
Comment (amrbashir) [inline tooling/cli/Cargo.toml:9]: Even though this is a CLI and a MSRV shouldn't matter that much but I think we plan to keep 1.60 for 1.x. Anyways, this PR probably needs to target `dev` not `1.x` so this shouldn't be a problem.

File: tooling/bundler/src/bundle/windows/sign.rs [hunk 2]
```diff
@@ -17,14 +19,164 @@ use winreg::{
...
-pub struct SignParams {
+/// Enum to hold the different signing params depending on the current setup. (Signtool or AzureSignTool)
+pub enum SignParams {
+  SignTool(SignToolParams),
+  Azure(AzureSignToolParams),
+}
+
+impl SignParams {
+  /// Locates the signing tool executable
+  fn locate_tool(&self) -> crate::Result<PathBuf> {
```
Comment (amrbashir) [inline tooling/bundler/src/bundle/windows/sign.rs]: Seems like is never used, let's remove for now

File: tooling/bundler/src/bundle/windows/msi/wix.rs [hunk 3]
```diff
@@ -788,8 +785,9 @@ pub fn build_wix_app_installer(
       &msi_output_path,
     )?;
     rename(&msi_output_path, &msi_path)?;
-    try_sign(&msi_path, settings)?;
-    output_paths.push(msi_path);
```
Comment (amrbashir) [inline tooling/bundler/src/bundle/windows/msi/wix.rs:792]: this line shouldn't be removed

File: tooling/bundler/src/bundle/windows/sign.rs [hunk 4]
```diff
@@ -88,93 +240,84 @@ fn locate_signtool() -> crate::Result<PathBuf> {
...
+    // We'll attempt to create a SignParams struct from the environment variables and the settings
+    // First we'll start with AzureSignTool, if we have all the required environment variables set, then we'll use that.
+    // If not, we'll fallback to Signtool.
+    // But if we don't have a certificate thumbprint set, we can't sign at all and we'll return None.
+    match (
+      var_os("AZURE_KEYVAULT_URL"),
+      var_os("AZURE_CLIENT_ID"),
+      var_os("AZURE_TENANT_ID"),
+      var_os("AZURE_CLIENT_SECRET"),
+      var_os("AZURE_CERTIFICATE_NAME"),
```
Comment (amrbashir) [inline tooling/bundler/src/bundle/windows/sign.rs:258]: we need to list all of these new env variables in `tooling/cli/ENVIRONMENT_VARIABLES.md`

**Review by amrbashir — CHANGES_REQUESTED** [review]
I am not sure if we should support `azuresigntool` tbh since it is not an official tool (correct me if I am wrong), should we instead provide a generic option to specify a signing command, similar to this https://github.com/crabnebula-dev/cargo-packager/blob/ec1a3a12296275341fd59c083d98202de316805f/crates/packager/src/config/mod.rs#L1206-L1214?

File: tooling/bundler/src/bundle/windows/sign.rs (hunk 4 above)
Comment (tr3ysmith) [reply tooling/bundler/src/bundle/windows/sign.rs:258]: Added!

File: tooling/bundler/src/bundle/windows/msi/wix.rs (hunk 3 above)
Comment (tr3ysmith) [reply tooling/bundler/src/bundle/windows/msi/wix.rs:792]: @amrbashir its still there, it just comes from sign_params now right below it. I added back `output_paths.push(msi_path);`

File: tooling/bundler/src/bundle/windows/sign.rs (hunk 2 above)
Comment (tr3ysmith) [reply tooling/bundler/src/bundle/windows/sign.rs]: removed!

File: tooling/cli/Cargo.toml (hunk 1 above)
Comment (tr3ysmith) [reply tooling/cli/Cargo.toml:9]: removed!

**Comment by tr3ysmith** [comment]
@amrbashir do I need to do anything to fix the clippy workflow failure?

**Comment by amrbashir** [comment]
if you want, that would be appreciated, otherwise we can fix it in another PR

**Comment by amrbashir** [comment]
cc @lucasfernog about https://github.com/tauri-apps/tauri/pull/8718#pullrequestreview-1870698873

**Comment by lucasfernog** [comment]
> I am not sure if we should support `azuresigntool` tbh since it is not an official tool (correct me if I am wrong), should we instead provide a generic option to specify a command to signing command, similar to this [crabnebula-dev/cargo-packager@`ec1a3a1`/crates/packager/src/config/mod.rs#L1206-L1214](https://github.com/crabnebula-dev/cargo-packager/blob/ec1a3a12296275341fd59c083d98202de316805f/crates/packager/src/config/mod.rs#L1206-L1214)?

It gets complicated when you think about how complicated it is to sign a Windows app.. we need a better solution, and this might be one (I still didn't run through it yet, hopefully when I get back from vacation).

**Comment by lucasfernog** [comment]
> @lucasfernog should we bump the rust min version on the main crate too? I had to bump it to 1.72 on the CLI crate because of the dependency for jsonschema -> aHash (aHash requires 1.72)
> 
> [tkaitchuck/aHash#196](https://github.com/tkaitchuck/aHash/pull/196)

Yeah we should.. we need to do a better job on the msrv for the CLI

**Comment by tr3ysmith** [comment]
> > @lucasfernog should we bump the rust min version on the main crate too? I had to bump it to 1.72 on the CLI crate because of the dependency for jsonschema -> aHash (aHash requires 1.72)
> > [tkaitchuck/aHash#196](https://github.com/tkaitchuck/aHash/pull/196)
> 
> Yeah we should.. we need to do a better job on the msrv for the CLI

@lucasfernog @amrbashir should we then bump both the CLI and Core, or just CLI?

**Comment by tr3ysmith** [comment]
> if you want, that would be appreciated, otherwise we can fix it in another PR

I think we should do it in another PR, since it looks to be completely unrelated to this change

**Comment by lucasfernog** [comment]
that's ok we can do that separately

**Comment by tr3ysmith** [comment]
@lucasfernog MSRV for cli is updated

**Comment by WalrusSoup** [comment]
> I am not sure if we should support `azuresigntool` tbh since it is not an official tool (correct me if I am wrong), should we instead provide a generic option to specify a signing command, similar to this https://github.com/crabnebula-dev/cargo-packager/blob/ec1a3a12296275341fd59c083d98202de316805f/crates/packager/src/config/mod.rs#L1206-L1214?

+1. We just got our new digicert token and it would be nice to be able to just call our own signing script. Otherwise, I could see this just growing and growing to accommodate the dozens of signing approaches.

**Comment by amrbashir** [comment]
I agree, @tr3ysmith would a generic command work for you as well?

**Comment by tr3ysmith** [comment]
@amrbashir sorry ya'll. I've had some crazy life stuff pop up. Yes a generic command would work for me as long as I can pass all the required arguments through via the tauri-action. I can work on that here in the next few days.

**Comment by amrbashir** [comment]
thanks @tr3ysmith hope you're doing great, let me know if I can help with anything

**Comment by amrbashir** [comment]
Hi @tr3ysmith I have opened https://github.com/tauri-apps/tauri/pull/9865 to implement the custom signing command, gonna close this PR now in favor of that one. Thanks for your contributions

**Comment by tr3ysmith** [comment]
@amrbashir no problem, is this something that'll get merge into 1.x at all?

**Comment by amrbashir** [comment]
there is a PR to backport it in #9902

**Pull request closed without merging**